### PR TITLE
Inferschema

### PR DESCRIPTION
**Pull request by faresismail96**
## Summary
Infers the schema of a dataframe
**Contributor @Rayan958**

**Related Issue: #83**

**PR Type: Feature**

**Status: WIP**

**Breaking change? No**

### Solution
Added an infer schema possibility. given a dataset and a fs savepath, we will read the file, determine seperator, type... generate a dataframe and use that to create a yaml file with all the Attributes, Schemas and Domain listed.

### Deploy Notes
example of usage, 
spark-submit target/scala-2.11/comet-assembly-0.1.jar infer-schema file:///path/to/file.extension /path/to/results.yml header
header is a Boolean, to be used in the case of DSV

### Remaining Todos
This should be done before merging this PR:
* [x] - add unit tests

## Contributor checklist:
Go over all the following points, and put an `x` in all the boxes that apply.
If you're unsure about any of these, don't hesitate to ask. We're here to help!
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/scala/com/ebiznext/comet/job/Main.scala [hunk 1]
```diff
@@ -59,6 +60,7 @@ object Main extends StrictLogging {
         |comet import
         |comet ingest datasetDomain datasetSchema datasetPath
         |comet index --domain domain --schema schema --timestamp {@timestamp|yyyy.MM.dd} --id type-id --mapping mapping --format parquet|json|json-array --dataset datasetPath --conf key=value,key=value,...
+        |comet infer-schema datasetPath savePath header (Optional)
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/Main.scala]: I would use 
comet infer-schema ---input datasetpath --output outputPath --with-header
or
comet infer-schema -i datasetpath -o outputPath -h

-I : required
-o: optional, console by default
-h: optional, no header by default

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    * @return the domain name
+    */
+  def getMode: String = "FILE"
+
+  /** Get extension file
+    *
+    * @param path : file path
+    * @return the file extension
+    */
+  def getExtensionFile(path: Path) = {
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: do we need this ?

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+      ""
+  }
+
+  /** Get format file
+    *
+    * @param datasetInit : created dataset without specifying format
+    * @param extension : extension file
+    * @return
+    */
+  def getFormatFile(datasetInit: Dataset[String], extension: String): String = {
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: extension should not be required

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    val firstLine = datasetInit.first()
+
+    firstLine.charAt(0).toString match {
+      case "{" => "JSON"
+      case "[" => "ARRAY_JSON"
+      case _ =>
+        if (extension.matches(".*sv$")) "DSV"
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: Soulb be CSV candidate whatever is the extension

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+  /** Get separator file
+    *
+    * @param datasetInit : created dataset without specifying format
+    * @return the file separator
+    */
+  def getSeparator(datasetInit: Dataset[String]) = {
+    session.sparkContext
+      .parallelize(datasetInit.take(10))
+      .map(x => x.replaceAll("[A-Za-z0-9 \"'()?!éèîàÀÉÈç+]", ""))
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: I would rather extract the candidate separators explicitly

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+      ._1
+      .toString
+  }
+
+  /** Get domain name
+    *
+    * @param path : file path
+    * @return the domain name
+    */
+  def getDomainName(path: Path): String = {
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: Nope

Comment (hayssams) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: Domain should be a parameter of the infer-schema

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+  def getDomainDirectoryName(path: Path): String = {
+    path.toString.replace(path.getName, "")
+  }
+
+  /** Get schema name
+    *
+    * @param path : file path
+    * @return the schema name
+    */
+  def getSchemaName(path: Path): String = {
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: nope

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+  }
+
+  /** Get separator file
+    *
+    * @param datasetInit : created dataset without specifying format
+    * @return the file separator
+    */
+  def getSeparator(datasetInit: Dataset[String]) = {
+    session.sparkContext
+      .parallelize(datasetInit.take(10))
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: nope.
You are taking 10 lines but not the first 10 lines

File: src/main/scala/com/ebiznext/comet/schema/handlers/InferSchemaHandler.scala [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+import com.ebiznext.comet.schema.model._
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.types.{ArrayType, StructType}
+
+class InferSchemaHandler(dataframe: DataFrame) {
+
+   def createAttributes(schema: StructType): List[Attribute] = {
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/schema/handlers/InferSchemaHandler.scala]: LGTM

File: src/main/scala/com/ebiznext/comet/schema/handlers/InferSchemaHandler.scala [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                   ack: Option[String] = None): Domain = {
+
+    Domain(name: String, directory: String, metadata: Option[Metadata], schemas: List[Schema], comment: Option[String], extensions: Option[List[String]], ack: Option[String])
+  }
+
+
+
+  def generateYaml(domain: Domain, savePath:String): Unit = {
+    val data: Domain = domain
+    val mapper: ObjectMapper = new ObjectMapper(new YAMLFactory)
```
Comment (hayssams) [inline src/main/scala/com/ebiznext/comet/schema/handlers/InferSchemaHandler.scala]: We have a global mapper. Simply use it. No need to duplicate it here

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 4 above)
Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: Ok. So, it's a .txt file, we treat this as a CSV file ? (get the separator etc...)

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 5 above)
Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: It's not the most present separator ?

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 6 above)
Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: So, it must be added in command line.

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 7 above)
Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: Schema name is a parameter that needs to be added in command line ?

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 8 above)
Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: **take** function doesn't return the n first lines of a dataset?

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 4 above)
Comment (hayssams) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: 👍 even if it has no extension. 
For JSON array, I would also check that the last char is a ']'
For json, I would check if the last char is a '}'

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 8 above)
Comment (hayssams) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: Yes the first 10 lines of the dataset but not of the file

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 5 above)
Comment (hayssams) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: I would not consider regular letter and digit as potential separator. Also I do not see the \t char as a candidate here ?

Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: With my regex, I remove letters, digits and some characters that can be used with string like "?", "!", escape etc ... to consider only characters like \t, ",", ":" etc ...

File: src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala (hunk 8 above)
Comment (Rayan958) [reply src/main/scala/com/ebiznext/comet/job/infer/InferSchemaJob.scala]: I see what you mean but in this context, I can choose any lines and not focus on the first lines of the file no ?

File: src/main/scala/com/ebiznext/comet/schema/handlers/InferSchemaHandler.scala (hunk 10 above)
Comment (faresismail96) [reply src/main/scala/com/ebiznext/comet/schema/handlers/InferSchemaHandler.scala]: ok, will use the one in main.

File: src/main/scala/com/ebiznext/comet/job/Main.scala (hunk 1 above)
Comment (faresismail96) [reply src/main/scala/com/ebiznext/comet/job/Main.scala]: used this format. 
comet infer-schema --domain domainName --schema schemaName --input datasetpath --output outputPath --with-header boolean 
with header being optional

**Pull request merged**